src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/math/RobustBrentSolver.java

```diff
@@ -0,0 +1,439 @@
+        SMALLEST_ROOT, LARGEST_ROOT
+    }
+
+    private final double relativeAccuracy, absoluteAccuracy, functionValueAccuracy;
+    private int evaluations = 0;
+
+    public RobustBrentSolver(final double relativeAccuracy, final double absoluteAccuracy,
+                             final double functionValueAccuracy) {
+        this.relativeAccuracy = relativeAccuracy;
+        this.absoluteAccuracy = absoluteAccuracy;
+        this.functionValueAccuracy = functionValueAccuracy;
+    }
+
+    /**
+     * The main solver routine
+     *
+     * @param maxEval maximum function evaluations
+     * @param objFunc the objective function to perform root finding on
+     * @param meritFunc a merit function to choose the best root (if multiple roots are found)
+     * @param min lower bound for the root
+     * @param max upper bound for the root
+     * @param numBisections number of bisections for "combing" the search interval
+     * @param depth the fractal depth of the search grid
+     * @return the root
+     * @throws TooManyEvaluationsException if too many calls are made to the objective function
+     * @throws NoBracketingException if the root could not be ultimately bracketed
+     */
+    public double solve(final int maxEval,
+                        @Nonnull final UnivariateFunction objFunc,
+                        @Nullable final UnivariateFunction meritFunc,
+                        @Nullable final MeritPolicy meritPolicy,
+                        final double min, final double max,
+                        final int numBisections, final int depth)
+            throws TooManyEvaluationsException, NoBracketingException {
+        Utils.validateArg(meritFunc != null || meritPolicy != null, "Either a merit function or a merit policy" +
+                        " must be provided");
+        Utils.nonNull(objFunc, "The objective function must be non-null");
+        ParamUtils.isPositive(maxEval, "Maximum function evaluations must be positive");
+        ParamUtils.isPositiveOrZero(max - min, "The search interval is ill-posed");
+        ParamUtils.isPositiveOrZero(numBisections, "Number of search interval bisections must be non-negative");
+        ParamUtils.isPositiveOrZero(depth, "The search depth must be non-negative");
+
+        evaluations = 0;
+
+        /* create a search grid */
+        final double[] xSearchGrid = createFractalSearchGrid(min, max, numBisections, depth);
+
+        /* evaluate the function on the search grid */
+        final double[] fSearchGrid = new double[xSearchGrid.length];
+        for (int k = 0; k < xSearchGrid.length; k++) {
+            fSearchGrid[k] = objFunc.value(xSearchGrid[k]);
+            evaluations++;
+        }
+
+        final List<Bracket> bracketsList = detectBrackets(xSearchGrid, fSearchGrid);
+        if (bracketsList.isEmpty()) {
+            throw new NoBracketingException(min, max, fSearchGrid[0], fSearchGrid[fSearchGrid.length-1]);
+        } else {
```

Comment:
ok, even though it breaks the semantics of nested mutually exclusive conditions in my opinion.